proposal/embedded-protocol-2.md

```diff
@@ -0,0 +1,173 @@
+  default field values as absent.
+
+* Move `CompileSuccess.loaded_urls` to `CompileResult.loaded_urls` so it's
+  available even when compilation fails.
+
+## Overview
+
+### Packet Structure
+
+Replace the last paragraph of the [embedded protocol overview] with:
+
+[embedded protocol overview]: ../spec/embedded-protocol.md#overview
+
+Each message in the embedded protocol is sent as a _packet_ which contains two
+values: an unsigned [varint] up to 32 bits long known as the "compilation ID",
+and a protocol buffer that contains the protobuf message. For streams (like
+standard input and output) that don't have built-in message boundaries, every
+packet must begin with another unsigned varint indicating the length in bytes of
+the remaining message (_including the compilation ID_). This matches the best
+practice described in [the protocol buffer documentation].
+
+Because JavaScript can't easily represent integers larger than 2^53 - 1, the
+length may be no more than 2^53 - 1. Because it's so unlikely that this will
+come up in practice, implementations are not required to verify it.
+
+[varint]: https://developers.google.com/protocol-buffers/docs/encoding#varints
+[the protocol buffer documentation]: https://developers.google.com/protocol-buffers/docs/techniques#streaming
+
+For a length-delimited stream, each packet has the following structure:
+
+```
+╔══════════╦══════════════════╗
+║ varint   ║ Length           ║
+╠══════════╬══════════════════╣
+║ varint   ║ Compilation ID   ║
+╠══════════╬══════════════════╣
+║ protobuf ║ Protobuf Message ║
+╚══════════╩══════════════════╝
+```
+
+## RPCs
+
+### Type Definitions
+
+Replace the following RPC type definitions:
+
+* *Requests* usually include a mandatory `uint32 id` field so that the other
+  endpoint can respond, except for `CompileRequest` which uses the [compilation
+  ID] as its ID. All request message types end in `Request`.
+
+  [compilation ID]: #packet-structure
+
+* *Responses* usually include a mandatory `uint32 id` field whose value must be
+  the same as their associated request's `id`, except for `CompileResponse`
+  which uses the compilation ID as its ID. All response message types begin with
+  the corresponding request name and end with `Response`.
+
+### ID Requirements
+
+Replace the paragraph that beings "Implementations must guarantee that they use
+a unique `id` for every request" with:
+
+Each endpoint must guarantee that each request's `id` doesn't match the `id` of
+any other outstanding request from that endpoint, although the same `id` may be
+used for an inbound request and an outbound request. The host must similarly
+guarantee that a `CompileRequest`'s [compilation ID] doesn't match the
+compilation ID of any other outstanding `CompileRequest`. The compiler must
+ensure that all outbound requests' compilation IDs match that of the
+`CompileRequest` that triggered its associated compilation.
```

Comment:
We should clarify the difference between compilation id and request id inside a compilation.

It is ok for compiler side requests to have overlapping id as long as they belongs to different compilation, and it would be unnecessarily difficult to synchronize different worker to make sure they don't user overlapping id. 

E.g.

Compilation 1 -> FunctionCallRequest 1 
Compilation 2 -> FunctionCallRequest 1

Even if FunctionCallRequest share the same id, we can route them correctly because the routing is based on compilation id.

Reply:
Of course, in a single threaded implementation like what we have today, making the request id under all compilations completely unique is not a problem at all. It's just that we should remove this constraint when we move to a multi-threaded model.

Reply:
In other worlds, in a single threaded implementation we have today, the outstanding request on compiler side can be tracked via tuple of `(compilation_id, request_id)` instead of just `request_id`, therefore the `request_id` does not need to be unique.